pkg/clustermesh/mcsapi/serviceimport_controller.go

```diff
@@ -0,0 +1,546 @@
+			return strings.Compare(a.Cluster, b.Cluster)
+		}
+		if a.ExportCreationTimestamp.Before(&b.ExportCreationTimestamp) {
+			return -1
+		}
+		return 1
+	})
+	return orderedSvcExports
+}
+
+func isSimilarPort(a, b mcsapiv1alpha1.ServicePort) bool {
+	return a.Port == b.Port && a.Protocol == b.Protocol
+}
+
+func findSimilarPort(port mcsapiv1alpha1.ServicePort, ports []portMerge) int {
+	for i, currPort := range ports {
+		if isSimilarPort(port, currPort.ServicePort) {
+			return i
+		}
+	}
+	return -1
+}
+
+// checkDuplicatedPortNameConflict check if port has a duplicated port name in portsByName.
+func checkDuplicatedPortNameConflict(port portMerge, portsByName map[string]portMerge) (foundPortByName bool, hasConflict bool, conflictMsg string) {
+	portByName, foundPortByName := portsByName[port.Name]
+	if foundPortByName && !isSimilarPort(port.ServicePort, portByName.ServicePort) {
+		hasConflict = true
+		return foundPortByName, hasConflict, fmt.Sprintf(
+			"Duplicated port name \"%s\". Using port definition \"%v\" from oldest service export in cluster \"%s\".",
+			port.Name, portByName, portByName.cluster,
+		)
+	}
+
+	return foundPortByName, hasConflict, ""
+}
+
+func checkPortConflict(port, olderPort portMerge) string {
+	if port.Name != olderPort.Name {
+		return fmt.Sprintf(
+			"Conflicting port name on \"%v\". Using the port name \"%s\" from oldest service export in cluster \"%s\".",
+			port, olderPort.Name, olderPort.cluster,
+		)
+	}
+	if ptr.Deref(port.AppProtocol, "") != ptr.Deref(olderPort.AppProtocol, "") {
+		return fmt.Sprintf(
+			"Conflicting appProtocol on \"%v\". Using the appProtocol \"%s\" from oldest service export in cluster \"%s\".",
+			port, ptr.Deref(olderPort.AppProtocol, ""), olderPort.cluster,
+		)
+	}
+	return ""
+}
+
+// mergePorts merge all the ports into a map while doing conflict resolution
+// with the oldest CreationTimestamp. It also return if it detects any conflict
+func mergePorts(svcExportByCluster operator.ServiceExportsByCluster) ([]portMerge, string) {
+	conflict := ""
+	orderedSvcExports := orderSvcExportByPriority(svcExportByCluster)
+	ports := []portMerge{}
+	portsByName := map[string]portMerge{}
+	for _, svcExport := range orderedSvcExports {
+		for _, port := range svcExport.Ports {
+			portMergeValue := portMerge{
+				ServicePort:             port,
+				cluster:                 svcExport.Cluster,
+				ExportCreationTimestamp: svcExport.ExportCreationTimestamp,
+			}
+
+			foundPortByName := true
+			if port.Name != "" {
+				conflictDuplicatedPortName := ""
+				hasConflict := false
+				foundPortByName, hasConflict, conflictDuplicatedPortName = checkDuplicatedPortNameConflict(portMergeValue, portsByName)
+				if conflict == "" {
+					conflict = conflictDuplicatedPortName
+				}
+				if hasConflict {
+					continue
+				}
+			}
+
+			portIndex := findSimilarPort(port, ports)
+			if portIndex == -1 {
+				if !foundPortByName {
```

Comment:
When could this be true? I'd expect this conflict case to have already been caught by the logic above, when checking for the duplicate port names (as the name does exist, but the port/proto do not match).

Reply:
I'm confused, did you commented the right line?

If yes the `foundPortByName` is so that we never override any port name that was previously set (so that we always keep the oldest one). And if it was about why this whole check exists I think I just answered in the previous thread as it's a bit related to your previous question as well.

Reply:
Yeah, my question was essentially when the `foundPortByName` check does apply, and I agree that it is related to the other thread. The only case I could think of was regarding the unnamed ports, but removing that one it should no longer be necessary AFAICT.